In [2]:
print("Welcome to python practice session.")

Welcome to python practice session.


In [6]:
first_number = 100
second_number = 200

addition = first_number + second_number
print(addition)
print(type(addition),id(first_number))
print(id(first_number),id(second_number))


300
<class 'int'> 10917664
10917664 10920864


**# id:**
Its means that where the data is stored that means memory location .So, if declear id with variable its shows that which location that variable is stored.

Example : print(id(first_number),id(second_number))
The result will be:10917664 10920864 So its a meomry location.

**## Type :** 
Type means which data type are having in variable like if we declare a = 10 and we print that type as print(type(a)).
It shows the result is: class int>


In [4]:
a = 30 
b = 30 

print(id(a),id(b))

10915424 10915424


**Object intering**

```
### if we defined a variable having same values .Python will not creating the two object for this variable.
In case, if we print id in that case through #print(id(a),id(b)) which is returns same memory location.Becuase it is not creating different object for this.Same location where a & b is values 10 is stored.And its both a & b is refering same values.

So, its called Object intering.
```




# Compile time constant

if the primitive types or String types defined as constant and the value is known at compile time.**Then Compiler what will do????**
The compiler replaces all the constant every places in the code with its value.
This is called compiler time constant.


Now **question** is **What is primitive types???....**

Actually, primitive is just defined data types such as integer,float,double,boolean ,string.All of those are defined data types because before the variable declare, we oviously defined the data types although  the variable types cannot be changed.In others language we defined data types with variable .

As example - int a =10;

> Here, we delcare a variable and its intger types.

**But,**In python is not same as others language.Because, python is completely object-oriented which is not statically typed.

For example: In python. we do,

> a = 10 
here , This is defined variable as a integer ,When value is out into the a as 10 its means variable a is integer.
if we want to found it in programming way what is that types of variable.?????

We will write : print (types(a)) .

Now **question** is  what is **non-primitive types?**

Non-primitive data types are called reference types because they refer to objects.

Basically, its simply called objects while coding times which is created rather than predefined.





In [6]:
a = "string"
b = "string"
a in b

True

String Intering

In [7]:
"string" in "bal"

False

In [8]:
"string" is "string"

True

In [9]:
"stri"+"ng" is "string"

True

In [12]:
s1 = "strin"
s2 = "string"
#s1 + "g" in s2
s1 + "g" is s2

False

Why ????It's not working.Not working!!!!Hanging my head......

Because ,  interpreter is probably interning compile-time constants but not the results of run-time expressions.

So , i am gonna try another way.

In [25]:
S1 = "strin";
s11 = S1 + "g"
s2 = "string"
print(s11)
s11 is  s2
from six.moves import intern 

intern(s11) is s2

string


True

By using intern you ensure that you never create two string objects that have the same value - when you request the creation of a second string object with the same value as an existing string object, you receive a reference to the pre-existing string object. This way, you are saving memory. Also, string objects comparison is now very efficient because it is carried out by comparing the memory addresses of the two string objects instead of their content.

# We can do another way.

In [28]:
a = 'strin'
b = intern(a + 'g')

c = 'string'
b is c

True